File: packages/scheduling-engine/src/overflow/auto-route-overflow.ts
import {
  ALT_UTIL_THRESHOLD, DAY_CAP, MAX_ADVANCE_DAYS, MAX_AUTO_MOVES, MAX_OVERFLOW_ITER, S0, S2,
} from '../constants.js';
import type { DecisionRegistry } from '../decisions/decision-registry.js';
import { scheduleAll } from '../scheduler/scheduler.js';
import type { AdvanceAction, Block, MoveAction } from '../types/blocks.js';
import type { ConstraintConfig } from '../types/constraints.js';
import { DEFAULT_CONSTRAINT_CONFIG } from '../types/constraints.js';
import type { DecisionEntry } from '../types/decisions.js';
import type { EMachine, EOp, ETool } from '../types/engine.js';
import type { FeasibilityReport } from '../types/infeasibility.js';
import type { ResourceTimeline } from '../types/failure.js';
import type { DispatchRule } from '../types/kpis.js';
import type { TwinValidationReport } from '../types/twin.js';
import type { WorkforceConfig } from '../types/workforce.js';
import { computeOtdDeliveryFailures } from './otd-delivery-failures.js';
import { capAnalysis, computeAdvancedEdd, sumOverflow } from './overflow-helpers.js';
import type { TierState } from './tier-types.js';
import { runTier2 } from './tier2-tardiness.js';
import { runTier3, tier3Diag } from './tier3-otd-delivery.js';

export interface AutoRouteOverflowInput {
  ops: EOp[];
  mSt: Record<string, string>;
  tSt: Record<string, string>;
  userMoves: MoveAction[];
  machines: EMachine[];
  toolMap: Record<string, ETool>;
  workdays: boolean[];
  nDays: number;
  workforceConfig?: WorkforceConfig;
  rule?: DispatchRule;
  supplyBoosts?: Map<string, { boost: number }>;
  thirdShift?: boolean;
  constraintConfig?: ConstraintConfig;
  machineTimelines?: Record<string, ResourceTimeline>;
  toolTimelines?: Record<string, ResourceTimeline>;
  twinValidationReport?: TwinValidationReport;
  dates?: string[];
  orderBased?: boolean;
  maxTier?: 1 | 2 | 3 | 4;
}

export interface AutoRouteOverflowResult {
  blocks: Block[];
  autoMoves: MoveAction[];
  autoAdvances: AdvanceAction[];
  decisions: DecisionEntry[];
  registry: DecisionRegistry;
  feasibilityReport: FeasibilityReport;
}
export function autoRouteOverflow(input: AutoRouteOverflowInput): AutoRouteOverflowResult {
  const {
    ops,
    mSt,
    tSt,
    userMoves,
    machines,
    toolMap,
    workdays,
    nDays,
    workforceConfig,
    rule = 'EDD',
    supplyBoosts,
    thirdShift,
    constraintConfig = DEFAULT_CONSTRAINT_CONFIG,
    machineTimelines,
    toolTimelines,
    twinValidationReport,
    dates,
    orderBased,
  } = input;
  const maxTier = input.maxTier ?? 4;

  const twinPartnerMap = new Map<string, string>();
  if (twinValidationReport?.twinGroups) {
    for (const tg of twinValidationReport.twinGroups) {
      twinPartnerMap.set(tg.opId1, tg.opId2);
      twinPartnerMap.set(tg.opId2, tg.opId1);
    }
  }

  const baseParams = {
    ops, mSt, tSt, machines, toolMap, workdays, nDays, workforceConfig,
    rule, supplyBoosts, thirdShift, constraintConfig, enforceDeadlines: false,
    machineTimelines, toolTimelines, twinValidationReport, dates, orderBased,
  };
  const runSchedule = (moves: MoveAction[], advances?: AdvanceAction[]) =>
    scheduleAll({ ...baseParams, moves, enableLeveling: false, advanceOverrides: advances });
  const runScheduleWithLeveling = (moves: MoveAction[], advances?: AdvanceAction[]) =>
    scheduleAll({ ...baseParams, moves, enableLeveling: true, advanceOverrides: advances });

  let schedResult = runSchedule(userMoves);
  let blocks = schedResult.blocks;
  let totalOverflowMin = sumOverflow(blocks);

  const autoMoves: MoveAction[] = [];
  const autoAdvances: AdvanceAction[] = [];

  const hardCap = thirdShift ? S2 - S0 : DAY_CAP;
  const preTier1TardyOps = new Set<string>();
  for (const b of blocks) {
    if (b.type === 'ok' && b.eddDay != null && b.dayIdx > b.eddDay) {
      preTier1TardyOps.add(b.opId);
    }
  }

  // ── TIER 1: Resolve OVERFLOW ──
  if (totalOverflowMin > 0) {
    const maxSteps = MAX_AUTO_MOVES * MAX_OVERFLOW_ITER;

    for (let step = 0; step < maxSteps; step++) {
      if (autoMoves.length + autoAdvances.length >= MAX_AUTO_MOVES) break;

      const cap = capAnalysis(blocks, machines);
      const actionIds = new Set([
        ...userMoves.map((m) => m.opId),
        ...autoMoves.map((m) => m.opId),
        ...autoAdvances.map((a) => a.opId),
      ]);

      const allOverflowBlocks = blocks
        .filter(
          (b) =>
            ((b.overflow && b.overflowMin != null && b.overflowMin > 0) ||
              (b.type === 'infeasible' && b.prodMin > 0)) &&
            !actionIds.has(b.opId),
        )
        .sort((a, b) => {
          const aMin = a.overflow ? a.overflowMin || 0 : a.prodMin;
          const bMin = b.overflow ? b.overflowMin || 0 : b.prodMin;
          return bMin - aMin;
        });

      if (allOverflowBlocks.length === 0) break;

      let moved = false;
      const seenOps = new Set<string>();

      for (const ob of allOverflowBlocks) {
        if (seenOps.has(ob.opId)) continue;
        seenOps.add(ob.opId);

        const mId = ob.machineId;
        const mDays = cap[mId];
        if (!mDays) continue;

        for (let advDays = 1; advDays <= MAX_ADVANCE_DAYS; advDays++) {
          const targetDay = computeAdvancedEdd(ob.dayIdx, advDays, workdays);
          if (targetDay < 0) break;
          const trial: AdvanceAction[] = [
            ...autoAdvances,
            { opId: ob.opId, advanceDays: advDays, originalEdd: ob.dayIdx },
          ];
          const newResult = runSchedule([...userMoves, ...autoMoves], trial);
          const newOverflow = sumOverflow(newResult.blocks);

          if (newOverflow < totalOverflowMin) {
            autoAdvances.push({ opId: ob.opId, advanceDays: advDays, originalEdd: ob.dayIdx });
            blocks = newResult.blocks;
            schedResult = newResult;
            totalOverflowMin = newOverflow;
            moved = true;

            for (const b of blocks) {
              if (b.opId === ob.opId && b.type === 'ok') {
                b.isAdvanced = true;
                b.advancedByDays = advDays;
              }
            }

            const unschedMin = ob.overflow ? ob.overflowMin || 0 : ob.prodMin;
            schedResult.registry.record({
              type: 'ADVANCE_PRODUCTION',
              opId: ob.opId,
              toolId: ob.toolId,
              machineId: mId,
              detail: `Advanced ${ob.sku} production by ${advDays} working days on ${mId} (EDD ${ob.dayIdx} -> ${targetDay}, ${ob.type === 'infeasible' ? 'infeasible' : 'overflow'}: ${unschedMin}min)`,
              metadata: {
                originalEdd: ob.dayIdx,
                newEdd: targetDay,
                advanceDays: advDays,
                sku: ob.sku,
                overflowMin: unschedMin,
                machineId: mId,
              },
            });

            break;
          }
        }
        if (moved) break;
      }

      // ── Phase B: Alt machine routing ──
      if (!moved) {
        const altCandidates = allOverflowBlocks.filter(
          (b) =>
            b.hasAlt &&
            b.altM &&
            !actionIds.has(b.opId) &&
            mSt[b.altM!] !== 'down' &&
            !(
              machineTimelines?.[b.altM!] &&
              machineTimelines[b.altM!].every((day) =>
                (thirdShift ? (['X', 'Y', 'Z'] as const) : (['X', 'Y'] as const)).every(
                  (s) => (day[s]?.capacityFactor ?? 1) <= 0,
                ),
              )
            ),
        );

        const seenOpsAlt = new Set<string>();
        for (const ob of altCandidates) {
          if (seenOpsAlt.has(ob.opId)) continue;
          seenOpsAlt.add(ob.opId);

          const altM = ob.altM!;
          const altDays = cap[altM];
          if (!altDays) continue;

          const wDayCount = workdays ? workdays.filter(Boolean).length : nDays;
          const altTotalUsed = altDays.reduce((s, d) => s + d.prod + d.setup, 0);
          const altUtil = altTotalUsed / (wDayCount * hardCap);
          if (altUtil > ALT_UTIL_THRESHOLD) continue;

          const altRemaining = wDayCount * hardCap - altTotalUsed;
          if (altRemaining < 30) continue;

          const twinPartner = twinPartnerMap.get(ob.opId);
          const twinAlreadyMoved = twinPartner ? actionIds.has(twinPartner) : true;
          autoMoves.push({ opId: ob.opId, toM: altM });
          if (twinPartner && !twinAlreadyMoved) {
            autoMoves.push({ opId: twinPartner, toM: altM });
          }
          const newResult = runSchedule(
            [...userMoves, ...autoMoves],
            autoAdvances.length > 0 ? autoAdvances : undefined,
          );
          const newOverflow = sumOverflow(newResult.blocks);

          if (newOverflow < totalOverflowMin) {
            blocks = newResult.blocks;
            schedResult = newResult;
            totalOverflowMin = newOverflow;
            moved = true;
            const unschedMin = ob.overflow ? ob.overflowMin || 0 : ob.prodMin;
            schedResult.registry.record({
              type: 'OVERFLOW_ROUTE',
              opId: ob.opId,
              toolId: ob.toolId,
              machineId: altM,
              detail: `Moved ${ob.sku} from ${ob.machineId} to ${altM} (${ob.type === 'infeasible' ? 'infeasible' : 'overflow'}: ${unschedMin}min)`,
              metadata: {
                fromMachine: ob.machineId,
                toMachine: altM,
                overflowMin: unschedMin,
                sku: ob.sku,
                altUtil: Math.round(altUtil * 100),
              },
            });

            break;
          } else {
            autoMoves.pop();
            if (twinPartner && !twinAlreadyMoved) autoMoves.pop();
          }
        }
      }

      if (!moved) break;
      if (totalOverflowMin === 0) break;
    }
  }

  // Save Tier 1 decisions before Tier 2/3 replace schedResult
  const tier1Decisions = schedResult.registry.getAll().filter(
    (d) => d.type === 'ADVANCE_PRODUCTION' || d.type === 'OVERFLOW_ROUTE',
  );

  const tierState: TierState = { blocks, schedResult, autoMoves, autoAdvances };
  const tierCtx = {
    ops,
    userMoves,
    mSt,
    workdays,
    twinPartnerMap,
    thirdShift,
    machineTimelines,
    runSchedule,
    runScheduleWithLeveling,
  };

  // ── TIER 2: Resolve TARDINESS ──
  if (maxTier >= 2) {
    runTier2(tierState, tierCtx, preTier1TardyOps);
  }

  if (maxTier >= 3) {
    const allRules: DispatchRule[] = ['EDD', 'ATCS', 'CR', 'SPT', 'WSPT'];
    let bestOtdDCount = computeOtdDeliveryFailures(tierState.blocks, ops).count;
    let bestRuleState: TierState | null = null;
    const baseSchedParams = {
      ops, mSt, tSt, machines, toolMap, workdays, nDays, workforceConfig,
      supplyBoosts, thirdShift, constraintConfig, machineTimelines, toolTimelines,
      twinValidationReport, dates, orderBased, enforceDeadlines: false,
    };

    for (const tryRule of allRules) {
      const mkSched = (lvl: boolean) => (moves: MoveAction[], advances?: AdvanceAction[]) =>
        scheduleAll({ ...baseSchedParams, rule: tryRule, enableLeveling: lvl, moves, advanceOverrides: advances });

      const ts: TierState = {
        blocks: tierState.blocks, schedResult: tierState.schedResult,
        autoMoves: [...tierState.autoMoves],
        autoAdvances: tierState.autoAdvances.map((a) => ({ ...a })),
      };
      const fresh = mkSched(false)([...userMoves, ...ts.autoMoves], ts.autoAdvances.length > 0 ? ts.autoAdvances : undefined);
      ts.blocks = fresh.blocks; ts.schedResult = fresh;

      runTier3(ts, { ...tierCtx, runSchedule: mkSched(false), runScheduleWithLeveling: mkSched(true) }, toolMap);
      const cnt = computeOtdDeliveryFailures(ts.blocks, ops).count;
      tier3Diag[`rule_${tryRule}`] = cnt;
      if (cnt < bestOtdDCount) { bestOtdDCount = cnt; bestRuleState = ts; }
      if (bestOtdDCount === 0) break;
    }

    if (bestRuleState) {
      tierState.blocks = bestRuleState.blocks;
      tierState.schedResult = bestRuleState.schedResult;
      tierState.autoMoves = bestRuleState.autoMoves;
      tierState.autoAdvances = bestRuleState.autoAdvances;
    }
  }

  // Collect all decisions from Tier 1 + current registry (Tier 2/3)
  const prevDecisions = [
    ...tier1Decisions,
    ...tierState.schedResult.registry.getAll().filter(
      (d) => d.type === 'ADVANCE_PRODUCTION' || d.type === 'OVERFLOW_ROUTE',
    ),
  ];
  const finalMoves = [...userMoves, ...tierState.autoMoves];
  const finalAdvances = tierState.autoAdvances.length > 0 ? tierState.autoAdvances : undefined;
  const commonParams = {
    ops, mSt, tSt, moves: finalMoves, machines, toolMap, workdays, nDays,
    workforceConfig, rule, supplyBoosts, thirdShift, constraintConfig,
    machineTimelines, toolTimelines, advanceOverrides: finalAdvances,
    twinValidationReport, dates, orderBased,
  };

  let finalResult = scheduleAll({ ...commonParams, enableLeveling: false, enforceDeadlines: false });
  let bestOtdD = computeOtdDeliveryFailures(finalResult.blocks, ops).count;

  for (const enableLeveling of [false, true]) {
    for (const enforceDeadlines of [false, true]) {
      if (!enableLeveling && !enforceDeadlines) continue;
      const candidate = scheduleAll({ ...commonParams, enableLeveling, enforceDeadlines });
      const candidateOtdD = computeOtdDeliveryFailures(candidate.blocks, ops).count;
      if (candidateOtdD < bestOtdD) {
        finalResult = candidate;
        bestOtdD = candidateOtdD;
      }
    }
  }

  // ── Post-grid OTD-D repair: boost failing ops and re-try with advance combos ──
  if (bestOtdD > 0 && bestOtdD <= 5) {
    const failResult = computeOtdDeliveryFailures(finalResult.blocks, ops);
    const failOps = new Set(failResult.failures.map((f) => f.opId));
    const boostedSupply = new Map(supplyBoosts ?? []);
    for (const opId of failOps) boostedSupply.set(opId, { boost: 10 });

    // Try with boosted supply + various targeted advance amounts for failing ops
    for (let advDays = 0; advDays <= 40; advDays++) {
      const extraAdvances: AdvanceAction[] = [];
      if (advDays > 0) {
        for (const f of failResult.failures) {
          const existing = tierState.autoAdvances.find((a) => a.opId === f.opId && (a.targetEdd === f.day || a.targetEdd == null))?.advanceDays ?? 0;
          const td = existing + advDays;
          if (computeAdvancedEdd(f.day, td, workdays) >= 0) {
            extraAdvances.push({ opId: f.opId, advanceDays: td, originalEdd: f.day, targetEdd: f.day });
          }
        }
      }
      const boostAdvances = finalAdvances ? [...finalAdvances] : [];
      for (const ea of extraAdvances) {
        const idx = boostAdvances.findIndex((a) => a.opId === ea.opId);
        if (idx >= 0) boostAdvances[idx] = ea; else boostAdvances.push(ea);
      }
      const boostParams = {
        ...commonParams,
        supplyBoosts: boostedSupply.size > 0 ? boostedSupply : undefined,
        advanceOverrides: boostAdvances.length > 0 ? boostAdvances : undefined,
      };

      for (const enableLeveling of [false, true]) {
        for (const enforceDeadlines of [false, true]) {
          const candidate = scheduleAll({ ...boostParams, enableLeveling, enforceDeadlines });
          const candidateOtdD = computeOtdDeliveryFailures(candidate.blocks, ops).count;
          if (candidateOtdD < bestOtdD) {
            finalResult = candidate;
            bestOtdD = candidateOtdD;
            if (bestOtdD === 0) break;
          }
        }
        if (bestOtdD === 0) break;
      }
      if (bestOtdD === 0) break;
    }
    // Update tierState advances if we added extras
    if (bestOtdD === 0) {
      for (const f of failResult.failures) {
        const extra = finalAdvances?.find((a) => a.opId === f.opId);
        if (extra && !tierState.autoAdvances.some((a) => a.opId === f.opId)) {
          tierState.autoAdvances.push(extra);
        }
      }
    }
  }

  const advMap = new Map(tierState.autoAdvances.map((a) => [a.opId, a.advanceDays]));
  for (const b of finalResult.blocks) {
    const advDays = advMap.get(b.opId);
    if (advDays != null && b.type === 'ok') {
      b.isAdvanced = true;
      b.advancedByDays = advDays;
    }
  }

  // Re-inject all Tier 1/2/3 advance/overflow decisions into final registry
  const existingIds = new Set(finalResult.registry.getAll().map((d) => d.opId + d.type));
  for (const dec of prevDecisions) {
    const key = dec.opId + dec.type;
    if (!existingIds.has(key)) {
      finalResult.registry.record(dec);
      existingIds.add(key);
    }
  }

  tierState.blocks = finalResult.blocks;
  tierState.schedResult = finalResult;

  return {
    blocks: tierState.blocks,
    autoMoves: tierState.autoMoves,
    autoAdvances: tierState.autoAdvances,
    decisions: tierState.schedResult.decisions,
    registry: tierState.schedResult.registry,
    feasibilityReport: tierState.schedResult.feasibilityReport,
  };
}

File: apps/backend/src/domain/scheduling/overflow/auto_route_overflow.py
"""Auto-route overflow — 3-tier overflow resolver.

Port of overflow/auto-route-overflow.ts (452 LOC).
Tier 1: resolve overflow (advance + alt machine routing).
Tier 2: resolve tardiness (advance + alt + combo + batch).
Tier 3: resolve OTD-delivery failures (multi-rule search).
"""

from __future__ import annotations

from ..constants import (
    ALT_UTIL_THRESHOLD,
    DAY_CAP,
    MAX_AUTO_MOVES,
    MAX_OVERFLOW_ITER,
    S0,
    S2,
)
from ..scheduler.otd_delivery import compute_otd_delivery_failures
from ..scheduler.pipeline import schedule_all
from ..types import (
    AdvanceAction,
    EMachine,
    EOp,
    ETool,
    MoveAction,
    ScheduleResult,
    TwinValidationReport,
    WorkforceConfig,
)
from .overflow_helpers import (
    cap_analysis,
    compute_advanced_edd,
    sum_overflow,
)
from .tier2_tardiness import run_tier2
from .tier3_otd_delivery import run_tier3
from .tier_types import TierContext, TierState


def auto_route_overflow(
    *,
    ops: list[EOp],
    m_st: dict[str, str],
    t_st: dict[str, str],
    user_moves: list[MoveAction],
    machines: list[EMachine],
    tool_map: dict[str, ETool],
    workdays: list[bool],
    n_days: int,
    workforce_config: WorkforceConfig | None = None,
    rule: str = "EDD",
    supply_boosts: dict[str, dict] | None = None,
    third_shift: bool = False,
    constraint_config: dict | None = None,
    twin_validation_report: TwinValidationReport | None = None,
    dates: list[str] | None = None,
    order_based: bool | None = None,
    max_tier: int = 4,
) -> dict:
    """Run 3-tier overflow resolution.

    Returns dict with: blocks, auto_moves, auto_advances, decisions, feasibility.
    """
    # Build twin partner map
    twin_partner_map: dict[str, str] = {}
    if twin_validation_report and twin_validation_report.twin_groups:
        for tg in twin_validation_report.twin_groups:
            twin_partner_map[tg.op_id1] = tg.op_id2
            twin_partner_map[tg.op_id2] = tg.op_id1

    base_params = dict(
        ops=ops,
        m_st=m_st,
        t_st=t_st,
        machines=machines,
        tool_map=tool_map,
        workdays=workdays,
        n_days=n_days,
        workforce_config=workforce_config,
        rule=rule,
        third_shift=third_shift,
        twin_validation_report=twin_validation_report,
        order_based=order_based if order_based is not None else False,
    )

    def run_schedule(
        moves: list[MoveAction],
        advances: list[AdvanceAction] | None = None,
    ) -> ScheduleResult:
        return schedule_all(
            **base_params,
            moves=moves,
            enable_leveling=False,
            advance_overrides=advances,
        )

    def run_schedule_with_leveling(
        moves: list[MoveAction],
        advances: list[AdvanceAction] | None = None,
    ) -> ScheduleResult:
        return schedule_all(
            **base_params,
            moves=moves,
            enable_leveling=True,
            advance_overrides=advances,
        )

    sched_result = run_schedule(user_moves)
    blocks = sched_result.blocks
    total_overflow_min = sum_overflow(blocks)

    auto_moves: list[MoveAction] = []
    auto_advances: list[AdvanceAction] = []

    hard_cap = S2 - S0 if third_shift else DAY_CAP

    pre_tier1_tardy_ops: set[str] = set()
    for b in blocks:
        if b.type == "ok" and b.edd_day is not None and b.day_idx > b.edd_day:
            pre_tier1_tardy_ops.add(b.op_id)

    # ═══ TIER 1: Resolve OVERFLOW ═══
    if total_overflow_min > 0:
        max_steps = MAX_AUTO_MOVES * MAX_OVERFLOW_ITER

        for _step in range(max_steps):
            if len(auto_moves) + len(auto_advances) >= MAX_AUTO_MOVES:
                break

            cap = cap_analysis(blocks, machines)
            action_ids = set(
                [m.op_id for m in user_moves]
                + [m.op_id for m in auto_moves]
                + [a.op_id for a in auto_advances]
            )

            all_overflow_blocks = sorted(
                [
                    b
                    for b in blocks
                    if (
                        (b.overflow and b.overflow_min is not None and b.overflow_min > 0)
                        or (b.type == "infeasible" and b.prod_min > 0)
                    )
                    and b.op_id not in action_ids
                ],
                key=lambda b: (b.overflow_min or 0) if b.overflow else b.prod_min,
                reverse=True,
            )

            if not all_overflow_blocks:
                break

            moved = False
            seen_ops: set[str] = set()

            # Phase A: Advance production
            for ob in all_overflow_blocks:
                if ob.op_id in seen_ops:
                    continue
                seen_ops.add(ob.op_id)

                m_id = ob.machine_id
                m_days = cap.get(m_id)
                if not m_days:
                    continue

                for adv_days in range(1, 200):
                    target_day = compute_advanced_edd(ob.day_idx, adv_days, workdays)
                    if target_day < 0:
                        break
                    trial = list(auto_advances) + [
                        AdvanceAction(
                            op_id=ob.op_id, advance_days=adv_days, original_edd=ob.day_idx
                        )
                    ]
                    new_result = run_schedule(list(user_moves) + list(auto_moves), trial)
                    new_overflow = sum_overflow(new_result.blocks)

                    if new_overflow < total_overflow_min:
                        auto_advances.append(
                            AdvanceAction(
                                op_id=ob.op_id, advance_days=adv_days, original_edd=ob.day_idx
                            )
                        )
                        blocks = new_result.blocks
                        sched_result = new_result
                        total_overflow_min = new_overflow
                        moved = True

                        for b in blocks:
                            if b.op_id == ob.op_id and b.type == "ok":
                                b.is_advanced = True
                                b.advanced_by_days = adv_days
                        break

                if moved:
                    break

            # Phase B: Alt machine routing
            if not moved:
                alt_candidates = [
                    b
                    for b in all_overflow_blocks
                    if b.has_alt
                    and b.alt_m
                    and b.op_id not in action_ids
                    and m_st.get(b.alt_m) != "down"
                ]

                seen_ops_alt: set[str] = set()
                for ob in alt_candidates:
                    if ob.op_id in seen_ops_alt:
                        continue
                    seen_ops_alt.add(ob.op_id)

                    alt_m = ob.alt_m
                    if not alt_m:
                        continue
                    alt_days = cap.get(alt_m)
                    if not alt_days:
                        continue

                    w_day_count = sum(1 for w in workdays if w) if workdays else n_days
                    alt_total_used = sum(d["prod"] + d["setup"] for d in alt_days)
                    alt_util = (
                        alt_total_used / (w_day_count * hard_cap)
                        if w_day_count * hard_cap > 0
                        else 1.0
                    )
                    if alt_util > ALT_UTIL_THRESHOLD:
                        continue

                    alt_remaining = w_day_count * hard_cap - alt_total_used
                    if alt_remaining < 30:
                        continue

                    twin_partner = twin_partner_map.get(ob.op_id)
                    twin_already_moved = twin_partner in action_ids if twin_partner else True

                    auto_moves.append(MoveAction(op_id=ob.op_id, to_m=alt_m))
                    if twin_partner and not twin_already_moved:
                        auto_moves.append(MoveAction(op_id=twin_partner, to_m=alt_m))

                    new_result = run_schedule(
                        list(user_moves) + list(auto_moves),
                        auto_advances if auto_advances else None,
                    )
                    new_overflow = sum_overflow(new_result.blocks)

                    if new_overflow < total_overflow_min:
                        blocks = new_result.blocks
                        sched_result = new_result
                        total_overflow_min = new_overflow
                        moved = True
                        break
                    else:
                        auto_moves.pop()
                        if twin_partner and not twin_already_moved:
                            auto_moves.pop()

            if not moved:
                break
            if total_overflow_min == 0:
                break

    # Build TierState for Tier 2/3
    tier_state = TierState(
        blocks=blocks,
        sched_result=sched_result,
        auto_moves=auto_moves,
        auto_advances=auto_advances,
    )
    tier_ctx = TierContext(
        ops=ops,
        user_moves=user_moves,
        m_st=m_st,
        workdays=workdays,
        twin_partner_map=twin_partner_map,
        third_shift=third_shift,
        run_schedule=run_schedule,
        run_schedule_with_leveling=run_schedule_with_leveling,
    )

    # ═══ TIER 2: Resolve TARDINESS ═══
    if max_tier >= 2:
        run_tier2(tier_state, tier_ctx, pre_tier1_tardy_ops)

    # ═══ TIER 3: OTD-Delivery via multi-rule search ═══
    if max_tier >= 3:
        all_rules = ["EDD", "ATCS", "CR", "SPT", "WSPT"]
        best_otd_d_count = compute_otd_delivery_failures(tier_state.blocks, ops)[0]
        best_rule_state: TierState | None = None

        for try_rule in all_rules:

            def mk_sched(lvl: bool, r: str = try_rule):
                def fn(
                    moves: list[MoveAction], advances: list[AdvanceAction] | None = None
                ) -> ScheduleResult:
                    return schedule_all(
                        **{**base_params, "rule": r},
                        moves=moves,
                        enable_leveling=lvl,
                        advance_overrides=advances,
                    )

                return fn

            ts = TierState(
                blocks=tier_state.blocks,
                sched_result=tier_state.sched_result,
                auto_moves=list(tier_state.auto_moves),
                auto_advances=[
                    AdvanceAction(
                        op_id=a.op_id, advance_days=a.advance_days, original_edd=a.original_edd
                    )
                    for a in tier_state.auto_advances
                ],
            )

            # Fresh schedule with this rule
            all_moves = list(user_moves) + list(ts.auto_moves)
            adv = ts.auto_advances if ts.auto_advances else None
            fresh = mk_sched(False)(all_moves, adv)
            ts.blocks = fresh.blocks
            ts.sched_result = fresh

            rule_ctx = TierContext(
                ops=ops,
                user_moves=user_moves,
                m_st=m_st,
                workdays=workdays,
                twin_partner_map=twin_partner_map,
                third_shift=third_shift,
                run_schedule=mk_sched(False),
                run_schedule_with_leveling=mk_sched(True),
            )
            run_tier3(ts, rule_ctx, tool_map)

            cnt = compute_otd_delivery_failures(ts.blocks, ops)[0]
            if cnt < best_otd_d_count:
                best_otd_d_count = cnt
                best_rule_state = ts
            if best_otd_d_count == 0:
                break

        if best_rule_state:
            tier_state.blocks = best_rule_state.blocks
            tier_state.sched_result = best_rule_state.sched_result
            tier_state.auto_moves = best_rule_state.auto_moves
            tier_state.auto_advances = best_rule_state.auto_advances

    # ═══ Final grid: leveling × deadlines ═══
    final_moves = list(user_moves) + list(tier_state.auto_moves)
    final_advances = tier_state.auto_advances if tier_state.auto_advances else None
    common_params = {
        **base_params,
        "moves": final_moves,
        "advance_overrides": final_advances,
    }

    final_result = schedule_all(
        **common_params, enable_leveling=False, enforce_deadlines_enabled=False
    )
    best_otd_d = compute_otd_delivery_failures(final_result.blocks, ops)[0]

    for enable_leveling in [False, True]:
        for enforce_dl in [False, True]:
            if not enable_leveling and not enforce_dl:
                continue
            candidate = schedule_all(
                **common_params,
                enable_leveling=enable_leveling,
                enforce_deadlines_enabled=enforce_dl,
            )
            candidate_otd_d = compute_otd_delivery_failures(candidate.blocks, ops)[0]
            if candidate_otd_d < best_otd_d:
                final_result = candidate
                best_otd_d = candidate_otd_d

    # ── Post-grid OTD-D repair ──
    if best_otd_d > 0 and best_otd_d <= 5:
        fail_count, fail_list = compute_otd_delivery_failures(final_result.blocks, ops)
        fail_ops = set(f.op_id for f in fail_list)
        boosted_supply = dict(supply_boosts or {})
        for op_id in fail_ops:
            boosted_supply[op_id] = {"boost": 10}

        for adv_days in range(0, 41):
            extra_advances: list[AdvanceAction] = []
            if adv_days > 0:
                for f in fail_list:
                    existing = 0
                    for a in tier_state.auto_advances:
                        if a.op_id == f.op_id:
                            existing = a.advance_days
                            break
                    td = existing + adv_days
                    if compute_advanced_edd(f.day, td, workdays) >= 0:
                        extra_advances.append(
                            AdvanceAction(op_id=f.op_id, advance_days=td, original_edd=f.day)
                        )

            boost_advances = list(final_advances) if final_advances else []
            for ea in extra_advances:
                idx = next((i for i, a in enumerate(boost_advances) if a.op_id == ea.op_id), -1)
                if idx >= 0:
                    boost_advances[idx] = ea
                else:
                    boost_advances.append(ea)

            boost_params = {
                **common_params,
                "advance_overrides": boost_advances if boost_advances else None,
            }

            for enable_leveling in [False, True]:
                for enforce_dl in [False, True]:
                    candidate = schedule_all(
                        **boost_params,
                        enable_leveling=enable_leveling,
                        enforce_deadlines_enabled=enforce_dl,
                    )
                    candidate_otd_d = compute_otd_delivery_failures(candidate.blocks, ops)[0]
                    if candidate_otd_d < best_otd_d:
                        final_result = candidate
                        best_otd_d = candidate_otd_d
                        if best_otd_d == 0:
                            break
                if best_otd_d == 0:
                    break
            if best_otd_d == 0:
                break

    # Mark advanced blocks in final result
    adv_map = {a.op_id: a.advance_days for a in tier_state.auto_advances}
    for b in final_result.blocks:
        adv_days_val = adv_map.get(b.op_id)
        if adv_days_val is not None and b.type == "ok":
            b.is_advanced = True
            b.advanced_by_days = adv_days_val

    return {
        "blocks": final_result.blocks,
        "auto_moves": tier_state.auto_moves,
        "auto_advances": tier_state.auto_advances,
        "decisions": final_result.decisions,
        "feasibility": final_result.feasibility,
    }
